### PR TITLE
Cleanup usage of plugins to test Logstash-core

### PR DESCRIPTION
**Pull request by purbon**
Nowadays we use some plugins to test the behaviour of the logstash-core, while this is an acceptable behaviour it should be required that plugins behave as expected to run the core test.

The current idea is, clean up:
- Current plugin usage for dummy/mock plugins where need.
- Rewrite test to they don't require this plugins to be used.

Like this we'll win isolation in the test, so easy to reproduce problems without the need to think if there is a bug, or an issue introduced by the plugins used in testing.

This is an initial approach to remove this issue, next tasks related to that would be:
- Wrap group knowledge on how each parts of logstash should behave, starting by configuration, validation (config mixing) and pipeline behaviour.
- This would help us write proper test that will work as a reference.
- With the reference we can start tracking a more deep test refactor as they should be.

### CODE REVIEW

File: Gemfile.jruby-1.9.lock [hunk 1]
```diff
@@ -46,7 +45,7 @@ GEM
       unf (>= 0.0.5, < 1.0.0)
     faraday (0.9.1)
       multipart-post (>= 1.2, < 3)
-    ffi (1.9.10-java)
+    ffi (1.9.8-java)
```
Comment (untergeek) [inline Gemfile.jruby-1.9.lock]: Interesting.  I hadn't heard about a regression.

**Comment by untergeek** [comment]
Why all the version number regressions in the gem versions?

**Comment by purbon** [comment]
@untergeek because this PR include https://github.com/elastic/logstash/pull/3951 basically we merged a wrong Gemfile in for master. This is the reason why you see what you see.

**Comment by purbon** [comment]
jenkins test this

**Comment by purbon** [comment]
please jenkins, test this.

**Comment by purbon** [comment]
@ph happy to get feedback from you man.

**Comment by ph** [comment]
I love the idea looking into it.

**Comment by purbon** [comment]
@ph thanks much appreciate it

**Comment by colinsurprenant** [comment]
is the strategy to actually re-implement all these plugins as mocks?

**Comment by purbon** [comment]
@colinsurprenant for some, like the generator I kinda like the idea, however there might be situations like the canceled events test where using multiline could be done in another way. I'm happy to remove as mocks as possible, feel free to point me out where they make no sense.

**Comment by colinsurprenant** [comment]
re-implementing plugins as mocks feels weird and will create duplicate code and eventually not mean much as these mocks are not the real plugins and these mock could be buggy rendering the spec useless. mocks should typically not contain logic, if it does, they should be tested themselves which becomes completely redundant. 

If we decide to move forward with a strategy of eliminating the need of plugins in the core specs, I think the specs should be refactored to completely not depend on plugins logic to run. the original tests should probably be moved into integration tests. 

For example, the cancel logic instead of being tested using a multiline plugin should be rewritten to only test the cancel logic expectations in itself and completely remove the need of a plugin, not even a multiline mock.

**Comment by andrewvc** [comment]
I agree with @colinsurprenant on the mocking issue. I think we should have a small list of 'golden' plugins for this purpose.

**Comment by ph** [comment]
When I said I like the idea, I haven't look at all of it.
This is what I think similar to @andrewvc and @colinsurprenant 

- Remove the plugins from the core spec.
- If a plugin is needed this should be an integration tests.
- We should rewrite the tests that need plugins, they are obviously testing too much.

**Comment by andrewvc** [comment]
Thinking through it a bit more, I'm thinking that we should either do something a bit closer to what @pere mentioned in terms of dummy plugins, or move some 'golden' plugins to logstash core. 

I'm strongly opposed to these plugins being mocks, I would prefer that they be actual running plugins that do something, just that they live in the same source tree.

Ideally we'd have have 3 plugins, an input (a simple version of the generator plugin), a filter (maybe a super simple one that just adds a tag, and an output (maybe one that just counts received messages?). 

I think that sits at the right abstraction level in terms of being useful and 'real world', without us writing a big chunk of real code that users depend on.

**Comment by colinsurprenant** [comment]
- I agree with the idea that core tests should not depend on plugins.
- all tests that use plugins to test core functionalities like using the multiline plugin to test the cancellation logic should be rewritten to not depend on a plugin. 
- the only place where a dummy plugin should be required is for the plugins discovery/loading code.
- all tests that uses plugins to test core functionalities should be moved as "integration" tests. These tests are still "valid" but don't belong in the core tests.
- as we ramp up better core tests, we will still depend on the "integrations" tests being run to test core.

**Comment by purbon** [comment]
In the benefit of keeping this a sane discussion I would recommend we focus discussions in the issue, check #3948 for that.

Just a matter of context, I start this as an initial approach to have a non plugin dependant tests for core, by using mocks, something that I still believe is not a bad thing. This mocks used here are very simple and easy to understand. Although I do see here some strong opinions about following this path for this solution.

I would rather keep this open, but start also doing work on refactoring the test as requested here, just to be aware:

* this means a lot more work and per se means this problem is staying here longer
* the new tests could be not easy to understand as they testing the same thing by reviewers, requiring then double effort from them.

**Comment by purbon** [comment]
@andrewvc using running plugins makes no sense, the problem is here because we've a dependency with a third party component, what exactly mocks/stubs are for.

**Comment by purbon** [comment]
@ph the plugins are needed because the test were poorly written, but they are not integration test at all. They are more functional tests.

**Comment by andrewvc** [comment]
I don't think I did a good job of communicating clearly.

@purbon what I'm trying to say is that our integration tests shouldn't be on things that are gem dependencies (like real world plugins). But rather lightweight plugins that are part of the core to prevent issues when we bump version numbers. We need a way to test the pipeline without using 'real' plugins, but while using something with a bit more reality than a mock.

I agree that we should use mocks in places where we are testing unit behavior, as @colinsurprenant has pointed out.

**Comment by purbon** [comment]
@andrewvc so what you ask for is to move the code I made for this plugins, named mocks because they mock a real plugin to a gem? then is ok for you? The plugins I wrote here are lightweight plugins that do a simple work, give them another name if mock is not appropriate.

The actual problem here the test uses plugins a lot to do it work, see the sample method. and before going for a full refactor I started this as an initial step. And they are not integration test, they test behaviour, but using plugins (what is wrong, but we could use this lightweight dummy plugins). Is my idea more clear now?

**Comment by jordansissel** [comment]
@purbon +1 on having dummy plugins used for testing core behaviors where needed. I agree 'mock' isn't the right word, but reading https://github.com/purbon/logstash/blob/feature/cleanup_plugin_test_deps/spec/support/plugins_helper.rb makes it clear the purpose (some fixture classes used during testing). +1 :)

**Comment by purbon** [comment]
agree naming is not probably appropriate, my bad on this.

**Comment by jordansissel** [comment]
@purbon hah, no worries. I confess that I never remember the definitions of mocks/doubles/fakes/stubs/etc. :)

**Comment by andrewvc** [comment]
@purbon I really have to apologize. I mixed up this thread with something else and thought I was commenting on a design issue instead of a PR. I've got too many balls in the air doing refactors and have been moving too fast. I apologize for not devoting more focus to reading this PR.

I reviewed all the code. I love the approach, and think it all looks great. I hereby retract all my comments.

**Comment by ph** [comment]
@purbon your comment made me realize this.

This PR improve the situation *right now* which make sure we can run the core test without having external dependency, I am OK with it.

**Comment by colinsurprenant** [comment]
For the record, I still disagree with the approach. It removes/weakens still valid tests which could have been simply moved into integration tests, replaces plugins with dummy/fake/mock plugins which is rather useless and redundant code, does not address the problems properly which would be to rewrite the tests to correctly test the expected core behaviour without a plugin harness.

I am at peace with my disagreement and seems to not be able to have a convincing argument so please feel free to reach consensus without me.

**Comment by guyboertje** [comment]
I think this approach is a good first step on a long journey.  Basically, if creating correct tests for a codebase is hard then that usually means the code needs refactoring.

@purbon - consider stubbing the `Logstash::Plugin.lookup` method to return a suitable Mock rather than rewriting the config strings or perhaps a combination of both.

On the subject of Unit vs Integration tests - imagine unit testing a piece of code that interacts with an external HTTP API, one would mock the external server and code it (with suitable stimuli) to return various 2XX, 3XX, 4XX, 5XX status codes so one can verify that one's code reacts correctly.  Later, for integration testing, if the external server had test endpoints, you would use that.  Stripe has a great test system, if you use certain credit card numbers in your call, their end will respond with an associated error (e.g. Insufficient funds) which allows one to verify that all the edge cases are covered.
Why did I mention Stripe? Because, in LS core, to correctly verify some core behaviours the dependencies albeit real or mocked have to react in a predicable way.

Having the odious task of maintaining both the Plugins and their Mocks should provide ample incentive to refactor for minimum burden.

**Comment by purbon** [comment]
I like this idea @guyboertje,

> consider stubbing the Logstash::Plugin.lookup method to return a suitable Mock rather than rewriting > the config strings or perhaps a combination of both

It would have been awesome to get this ideas in an issue like #3948 but I will try to resume them there for posterity :+1: 

- cheers

**Comment by purbon** [comment]
removing the stalled tag we reach consensus about the approach. Going to rename the dummy plugins to have a more descriptive name.

**Comment by ph** [comment]
@purbon its that ready to review?

**Comment by ph** [comment]
@purbon Not sure of the state of this PR, I need rebasing feel free to reassign it to me went its ready.

**Comment by suyograo** [comment]
Closing for now. We need to revisit this strategy.
